File: bench/lib/bench.exs
# elixir benchmark harness -- mirrors bench/bench.l.
# Bench.run(name, work) auto-scales the repetition count (doubling until the run
# clears @min_ms), then times that count ONCE MORE and reports the second run --
# the scaling runs pay the fixed per-process costs (JIT warm-up, the heap's first
# growth), which otherwise ride on whichever power of two the doubling landed on.
# one line per bench, matching the other harnesses:
#     <name> <lang> <reps> <ms> <checksum>
# work is a 0-arity fun returning a deterministic checksum. the column label
# comes from BENCH_LANG (default "elixir").
defmodule Bench do
  @min_ms 200.0
  @lang System.get_env("BENCH_LANG") || "elixir"

  def run(name, work), do: loop(name, work, 1)

  defp loop(name, work, reps) do
    {ms, _chk} = window(work, reps)
    if ms >= @min_ms do
      {ms, chk} = window(work, reps)                  # warm
      IO.puts("#{name} #{@lang} #{reps} #{:erlang.float_to_binary(ms, decimals: 3)} #{chk}")
    else
      loop(name, work, reps * 2)
    end
  end

  defp window(work, reps) do
    t0 = System.monotonic_time(:nanosecond)
    chk = rep(work, reps, nil)
    {(System.monotonic_time(:nanosecond) - t0) / 1.0e6, chk}
  end

  defp rep(_work, 0, chk), do: chk
  defp rep(work, n, _), do: rep(work, n - 1, work.())
end

File: bench/lib/bench.jl
# julia benchmark harness -- mirrors bench/bench.l and lib/bench.py.
# bench(name, work) auto-scales the repetition count (doubling until the run
# clears MIN_MS), then times that count ONCE MORE and reports the second run --
# the scaling runs pay the fixed per-process costs (JIT warm-up, the heap's first
# growth), which otherwise ride on whichever power of two the doubling landed on.
# one line per bench, matching the other harnesses:
#     <name> <lang> <reps> <ms> <checksum>
# work is a nullary closure returning a deterministic checksum. BENCH_LANG sets
# the column label, default "julia". A single untimed warm-up call triggers
# Julia's JIT compilation up front, so the timed loop measures steady state
# (the per-function compile is excluded, like every other harness excludes
# interpreter/JIT startup).
using Printf

const MIN_MS = 200.0
const LANG = get(ENV, "BENCH_LANG", "julia")

function bench(name, work)
    work()                       # warm up: force JIT compilation before timing
    # Opaque (::Any) handle: calling through it is a dynamic dispatch the
    # optimizer can't see into, so it can't hoist or constant-fold a pure
    # `work` out of the timing loop (which would spin the rep-doubling forever).
    w = Base.inferencebarrier(work)
    run = reps -> begin
        t0 = time_ns()
        chk = nothing
        for _ in 1:reps
            chk = w()
        end
        ((time_ns() - t0) / 1.0e6, chk)
    end
    reps = 1
    while true
        (ms, chk) = run(reps)
        if ms >= MIN_MS
            (ms, chk) = run(reps)            # warm
            println("$name $LANG $reps $(@sprintf("%.3f", ms)) $chk")
            return
        end
        reps *= 2
    end
end
